plugins/workspace-plugin/src/workspace-folder-updater.ts

```diff
@@ -0,0 +1,91 @@
+/**********************************************************************
+ * Copyright (c) 2021 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+
+import * as theia from '@theia/plugin';
+const UPDATE_WORKSPACE_FOLDER_TIMEOUT = 5000;
+
+export class WorkspaceFolderUpdater {
+  private pendingFolders: string[] = [];
+  private addingWorkspaceFolderPromise: Promise<void> | undefined;
+
+  async addWorkspaceFolder(path: string): Promise<void> {
```

Comment:
Maybe I'm missing somethings, but couldn't we replace this method with this:
```
  private addingWorkspaceFolderPromise: Promise<void> = Promise.resolve();

  async addWorkspaceFolder(path: string): Promise<void> {
    const workspaceFolderPath = this.toValidWorkspaceFolderPath(path);

    return this.addingWorkspaceFolderPromise.then(()=>this.addFolder(workspaceFolderPath));
  }
```

Reply:
`updateWorkspaceFolders()` is [too tricky](https://github.com/eclipse-theia/theia/blob/4e1c1d429226280e6f0eff49658c733a38b50c4a/packages/plugin/src/theia.d.ts#L5428).

Adding a workspace folder using API without waiting for `onDidChangeWorkspaceFolders` leads to bugs - I played a lot with it and faced with bugs related to `Project tree` and so on.

Let's consider the case:
- we are adding a workspace folder `111` using API
- at the same time a request for adding `222` is coming`
- 300 milliseconds later - request to add `333`
- 500 milliseconds later - request to add `444`

The API works well if we add folders step by step, waiting for `onDidChangeWorkspaceFolders` is the required condition for each step.

Reply:
We could avoid all this complex logic by adding all workspace folders in one API request when all projects have cloned.
But it leads to bad UX - `Project tree` will be empty all time while projects are cloned. It's too critical when at least one project is big enough and cloning process takes a lot of time...

Reply:
But we would only start executing the addFolder promise for `222` once the promise for '111` is resolved. Doing 
```
await promise;
doStuff();
```
is strictl the same as doing 
```
promise.then(()=>doStuff()),
```

Reply:
Sorry, I didn't catch the idea - what about `333` and `444` at the same time?

Reply:
You're trying to delay the processing of folder 222 until after the processing of folder 111 has finished, right? But if you do 

processFolder111Promise.then(processFolder222Promise) that is exactly the same as when you write

```
await processFolder111Promise;
startProcessFolder222Promise;
```
It literally compiles to the same javascript. The try catch can also be handled like so: https://stackoverflow.com/questions/36596925/promise-ignore-catch-and-return-to-chain

In the end, both versions work, just that mine would be easier to understand and shorter, IMO.

Reply:
I see, thank you!
Yes, I understand that I could use something like `promise.then(()=>doStuff()),`

I thought your way allows to avoid keeping a set with pending folders for adding...
So, sorry - first I didn't catch your idea. Now I see what it was about...

Reply:
I think you acutally _can_ get rid of that set: just check the workspaces array whether the folder you're adding is in it before executing the `updateWorkspace`. The code is executed only a couple of times and the workspace folder is usually 2-3 entries at most, so I would choose clarity over performance here.

Reply:
You might have a couple of extra thenables on the stack, but who cares.